mlflow/genai/utils/trace_utils.py

```diff
@@ -394,32 +394,200 @@ def _to_dict(obj: Any) -> dict[str, Any]:
+               If None, uses the default model.
+
+    Returns:
+        Dictionary mapping span IDs to lists of retrieved chunks.
+        Each chunk is a dict with 'content' (str) and optional 'doc_uri' (str).
+        Returns empty dict if extraction fails or no retrieval contexts found.
+
+    Note:
+        This function does not raise exceptions. If extraction fails, returns empty dict.
     """
     if trace is None or trace.data is None:
         return {}
 
     top_level_retrieval_spans = _get_top_level_retrieval_spans(trace)
-    if len(top_level_retrieval_spans) == 0:
+    if len(top_level_retrieval_spans) > 0:
+        retrieved = {}
+        for retrieval_span in top_level_retrieval_spans:
+            try:
+                contexts = [_parse_chunk(chunk) for chunk in retrieval_span.outputs or []]
+                retrieved[retrieval_span.span_id] = [c for c in contexts if c is not None]
+            except Exception as e:
+                _logger.debug(
+                    f"Fail to get retrieval context from span: {retrieval_span}. Error: {e!r}"
+                )
+
+        if retrieved:
+            return retrieved
+
+    return _try_extract_retrieval_context_with_llm(trace, model)
+
+
+def _try_extract_retrieval_context_with_llm(
+    trace: Trace,
+    model: str | None = None,
+) -> dict[str, list[dict[str, Any]]]:
+    """
+    Use LLM with tools to extract retrieval context from trace.
+
+    This is a fallback when programmatic extraction fails. The LLM uses tool calling
+    to examine trace spans and locate retrieval contexts.
+
+    Args:
+        trace: The MLflow trace object to analyze.
+        model: Optional model URI to use for extraction. If None, uses default model.
+
+    Returns:
+        Dictionary mapping span IDs to lists of retrieved chunks, in the same format
+        as extract_retrieval_context_from_trace.
+    """
+    from mlflow.genai.judges.constants import _DATABRICKS_DEFAULT_JUDGE_MODEL
+    from mlflow.genai.judges.utils import (
+        get_chat_completions_with_structured_output,
+        get_default_model,
+    )
+    from mlflow.types.llm import ChatMessage
+
+    output_example = json.dumps(
+        RetrievedChunksForTrace(
+            retrieval_contexts=[
+                RetrievedChunksForSpan(
+                    span_id="span ID where data was retrieved",
+                    chunks=[
+                        RetrievedChunk(
+                            content="text content as string (required)",
+                            doc_uri="source URI if available, otherwise null",
+                        )
+                    ],
+                )
+            ]
+        ).model_dump(),
+        indent=2,
+    )
+
+    model_uri = model or get_default_model()
+
+    # Skip LLM extraction if using Databricks default model
+    #
+    # TODO: Implement client logic for calling Databricks default model
+    # with tools and structured output for information extraction
```

Comment:
Existing methods for calling the Databricks default judge model with tools always return `Feedback`. We need some follow-up work to support the Databricks default judge model in `get_chat_completions_with_structured_output`